### ISSUE TITLE
Module not found: Can't resolve 'fs' in 'node_modules\electron'

### ISSUE THREAD

**Issue by Manuel-Rw** [electron-react-boilerplate/electron-react-boilerplate#2961]
## Summary
I am experiencing issues importing packages. Here, i imported ipc using ``import { ipcRenderer } from 'electron'``.
However, it doesn't matter *what* i import - it occurs with every package i use in React components.
```
 × ｢wdm｣: ERROR in ./node_modules/electron/index.js 1:11-24
Module not found: Error: Can't resolve 'fs' in 'E:\node_apps\chat-app\chat-app\node_modules\electron'
 @ ./src/renderer/components/Titlebar.tsx 9:22-53
 @ ./src/renderer/App.tsx 11:0-45 29:38-46
 @ ./src/renderer/index.tsx 7:0-24 8:41-44
 ```

I'm pretty new to Electron, but not to React (maybe i'm overlooking something?).
[This post didn't help me, but might be related.](https://stackoverflow.com/questions/48476061/electron-and-typescript-fs-cant-be-resolved)

## My environment
**node:** 16.8.0
**npm:** 7.21.0
**yarn**: 1.22.11

**Comment by evanandrewrose** [comment]
https://github.com/electron-react-boilerplate/electron-react-boilerplate/issues/2949

https://electron-react-boilerplate.js.org/docs/native-modules/

**Comment by evanandrewrose** [comment]
@Manuel-Rw 

I was struggling with the same thing and I originally thought the solution I linked was the correct approach (first day on electron). But I'm now understanding that IPC is designed to solve the security problem that we're evading with the link I shared.

Anyway, take a look at the IPC event emitted in `index.ejs`. It's using the API exposed via `preload.js`, which I think is a common pattern for electron apps. This interface forwards requests to the event handler in Main.ts. By modifying those touch points, I was able to get my fs-related API call working from my react components.

So, you shouldn't need to import ipcRenderer directly into your renderer `.tsx` files. Instead, you can use the API you manage via `window.electron.ipcRenderer`, as is done in the boilerplate ipc example.

**Comment by Manuel-Rw** [comment]
Thanks for your comment.
While it does make sense reading it, could you give me a real world example?
I'm quite confused, since i can't find any ``window.electron.ipcRenderer`` object in my ``.tsx`` components.
Also, the built-in example of IPC is not used in any ``tsx`` component but in a ``ejs`` file.

My goal is to close, minimize and maximize the electron window from a react component.

**Comment by wayaway-jtm** [comment]
> Thanks for your comment.
> While it does make sense reading it, could you give me a real world example?
> I'm quite confused, since i can't find any `window.electron.ipcRenderer` object in my `.tsx` components.
> Also, the built-in example of IPC is not used in any `tsx` component but in a `ejs` file.
> 
> My goal is to close, minimize and maximize the electron window from a react component.

Try [this answer](https://github.com/electron-react-boilerplate/electron-react-boilerplate/issues/2955#issuecomment-907397111).

Basically the problem is Typescript doesn't recognize `electron` as a property of `window` so it errors out during the build process. If you add the `global.d.ts` file like in the link above it basically says "hey Typescript `electron` does exist on `window`".

Anything exposed in the `preload.js` via `contextBridge.exposeInMainWorld` will also have to be added to the `global.d.ts` file. The reason for this is security. This way the browser application only has access to whatever properties it needs instead of the entirety of `ipcRenderer` or node or whatever is exposed.

**Comment by alielmajdaoui** [comment]
> @manuel-rw
> 
> I was struggling with the same thing and I originally thought the solution I linked was the correct approach (first day on electron). But I'm now understanding that IPC is designed to solve the security problem that we're introducing with the link I shared.
> 
> Take a look at the IPC event emitted in `index.ejs`. It's using the API exposed via `preload.js`, which I think is a common pattern for electron apps. This interface forwards requests to the event handler in Main.ts. By modifying those touch points, I was able to get my fs-related API call working from my react components.
> 
> So, you shouldn't need to import ipcRenderer directly into your renderer `.tsx` files. Instead, you can use the API you manage via `window.electron.ipcRenderer`, as is done in the boilerplate ipc example.

TLDR; if you have an import of `ipcRenderer` from react library on any renderer file, the error will occur!

**Comment by arnaudambro** [comment]
> #2949
> 
> https://electron-react-boilerplate.js.org/docs/native-modules/

I did that, but I can see it's not the first thing to do.

The first thing to do is: do NOT import directly `ipcRenderer` in any front-end javascript files.

in [./src/main/preload.ts](https://github.com/electron-react-boilerplate/electron-react-boilerplate/blob/main/src/main/preload.ts) there is 

https://github.com/electron-react-boilerplate/electron-react-boilerplate/blob/944dad54e31c53f84db7215fa9db7487ceba32b3/src/main/preload.ts#L27

then you can 
- add any ipcRenderer method you want to that `electronHandler` object
- use them with `window.electron.ipcRenderer`

and job done

**Comment by sebastianripari** [comment]
Hi all !

For all interested in use **ipcRenderer**, here is a example how to use it:
https://github.com/electron-react-boilerplate/electron-react-boilerplate/blob/v4.6.0/src/renderer/index.tsx

`// calling IPC exposed from preload script
window.electron.ipcRenderer.once('ipc-example', (arg) => {
  // eslint-disable-next-line no-console
  console.log(arg);
});
window.electron.ipcRenderer.sendMessage('ipc-example', ['ping']);
`

**Comment by parvans** [comment]
ERROR in ./node_modules/electron/index.js 1:11-24
Module not found: Error: Can't resolve 'fs' in 'D:\work\CODECHAT\CodChat\node_modules\electron'

i got the same error 
please help me 🥲🥲

**Comment by Karanbhuva18** [comment]
i also face this issue 

ERROR in ./node_modules/electron/index.js 1:11-24
Module not found: Error: Can't resolve 'fs' in 'D:\Electron.js\my-electron-react-app\node_modules\electron'
ERROR in ./node_modules/electron/index.js 2:13-28
Module not found: Error: Can't resolve 'path' in 'D:\Electron.js\my-electron-react-app\node_modules\electron'

BREAKING CHANGE: webpack < 5 used to include polyfills for node.js core modules by default.
This is no longer the case. Verify if you need this module and configure a polyfill for it.

If you want to include a polyfill, you need to:
	- add a fallback 'resolve.fallback: { "path": require.resolve("path-browserify") }'
	- install 'path-browserify'
If you don't want to include a polyfill, you can use an empty module like this:
	resolve.fallback: { "path": false }
	
i want to use ipcrender in 


import React, { useState } from 'react';
import { ipcRenderer } from 'electron'
const App = () => {
  

  const handleClick = () => {
    // Send a message to the main process
   ipcRenderer.send("msg","hello from render")
  };

  // Listen for response from main process
 

  return (
    <div>
      <button onClick={handleClick}>Ping Main Process</button>
      
    </div>
  );
};

export default App;
